### PR TITLE
Add a format rejections function

### LINKED ISSUE #9559: Could we shorten the solver rejections message?
The solver rejections message can be very long with a lot of repetition. Could we shorten it?

- What we have now with `cabal-install` at 8070ab14773484549c50a3df5568698cd33cbc4c

```
$ cabal build all --enable-tests --enable-benchmarks --dry-run
HEAD is now at 71c2494 Further improvements in parsing attachments.
Resolving dependencies...
Error: [Cabal-7107]
Could not resolve dependencies:
[__0] next goal: pandoc (user goal)
[__0] rejecting: pandoc-3.1.11, pandoc-3.1.10 (constraint from user target
requires ==3.1.9)
[__0] rejecting: pandoc-3.1.9 (constraint from project config
https://stackage.org/lts-21.7/cabal.config requires ==3.0.1)
[__0] rejecting: pandoc-3.1.8, pandoc-3.1.7, pandoc-3.1.6.2, pandoc-3.1.6.1,
pandoc-3.1.6, pandoc-3.1.5, pandoc-3.1.4, pandoc-3.1.3, pandoc-3.1.2,
pandoc-3.1.1, pandoc-3.1, pandoc-3.0.1, pandoc-3.0, pandoc-2.19.2,
pandoc-2.19.1, pandoc-2.19, pandoc-2.18, pandoc-2.17.1.1, pandoc-2.17.1,
pandoc-2.17.0.1, pandoc-2.17, pandoc-2.16.2, pandoc-2.16.1, pandoc-2.16,
pandoc-2.15, pandoc-2.14.2, pandoc-2.14.1, pandoc-2.14.0.3, pandoc-2.14.0.2,
pandoc-2.14.0.1, pandoc-2.14, pandoc-2.13, pandoc-2.12, pandoc-2.11.4,
pandoc-2.11.3.2, pandoc-2.11.3.1, pandoc-2.11.3, pandoc-2.11.2, pandoc-2.11.1.1,
pandoc-2.11.1, pandoc-2.11.0.4, pandoc-2.11.0.2, pandoc-2.11.0.1, pandoc-2.11,
pandoc-2.10.1, pandoc-2.10, pandoc-2.9.2.1, pandoc-2.9.2, pandoc-2.9.1.1,
pandoc-2.9.1, pandoc-2.9, pandoc-2.8.1, pandoc-2.8.0.1, pandoc-2.8,
pandoc-2.7.3, pandoc-2.7.2, pandoc-2.7.1, pandoc-2.7, pandoc-2.6, pandoc-2.5,
pandoc-2.4, pandoc-2.3.1, pandoc-2.3, pandoc-2.2.3.2, pandoc-2.2.3.1,
pandoc-2.2.3, pandoc-2.2.2.1, pandoc-2.2.2, pandoc-2.2.1, pandoc-2.2,
pandoc-2.1.3, pandoc-2.1.2, pandoc-2.1.1, pandoc-2.1, pandoc-2.0.6,
pandoc-2.0.5, pandoc-2.0.4, pandoc-2.0.3, pandoc-2.0.2, pandoc-2.0.1.1,
pandoc-2.0.1, pandoc-2.0.0.1, pandoc-2.0, pandoc-1.19.2.4, pandoc-1.19.2.3,
pandoc-1.19.2.2, pandoc-1.19.2.1, pandoc-1.19.2, pandoc-1.19.1, pandoc-1.19,
pandoc-1.18, pandoc-1.17.2, pandoc-1.17.1, pandoc-1.17.0.3, pandoc-1.17.0.2,
pandoc-1.17.0.1, pandoc-1.17, pandoc-1.16.0.2, pandoc-1.16.0.1, pandoc-1.16,
pandoc-1.15.2.1, pandoc-1.15.2, pandoc-1.15.1.1, pandoc-1.15.1, pandoc-1.15.0.6,
pandoc-1.15.0.5, pandoc-1.15.0.4, pandoc-1.15.0.3, pandoc-1.15.0.2,
pandoc-1.14.1, pandoc-1.13.2.1, pandoc-1.13.2, pandoc-1.13.1, pandoc-1.13.0.1,
pandoc-1.13, pandoc-1.12.4.2, pandoc-1.12.4, pandoc-1.12.3.3, pandoc-1.12.3.2,
pandoc-1.12.3.1, pandoc-1.12.3, pandoc-1.12.2.1, pandoc-1.12.2, pandoc-1.12.1,
pandoc-1.12.0.2, pandoc-1.12.0.1, pandoc-1.12, pandoc-1.11.1, pandoc-1.11,
pandoc-1.10.1, pandoc-1.10.0.5, pandoc-1.10.0.4, pandoc-1.10.0.3,
pandoc-1.10.0.2, pandoc-1.10.0.1, pandoc-1.10, pandoc-1.9.4.5, pandoc-1.9.4.4,
pandoc-1.9.4.3, pandoc-1.9.4.2, pandoc-1.9.4.1, pandoc-1.9.4, pandoc-1.9.3,
pandoc-1.9.2, pandoc-1.9.1.2, pandoc-1.9.1.1, pandoc-1.9.1, pandoc-1.9.0.5,
pandoc-1.9.0.4, pandoc-1.9.0.3, pandoc-1.9.0.2, pandoc-1.9, pandoc-1.8.2.1,
pandoc-1.8.2, pandoc-1.8.1.2, pandoc-1.8.1.1, pandoc-1.8.1, pandoc-1.8.0.3,
pandoc-1.8.0.2, pandoc-1.8.0.1, pandoc-1.8, pandoc-1.6.0.1, pandoc-1.6,
pandoc-1.5.1.1, pandoc-1.5.1, pandoc-1.5.0.1, pandoc-1.5, pandoc-1.4,
pandoc-1.3, pandoc-1.2.1, pandoc-1.2, pandoc-1.1, pandoc-1.0.0.1, pandoc-1.0,
pandoc-0.46, pandoc-0.45, pandoc-0.44, pandoc-0.43, pandoc-0.42, pandoc-0.41,
pandoc-0.4, pandoc-2.11.0.3, pandoc-1.15.0.1, pandoc-1.15, pandoc-1.14.0.4,
pandoc-1.14.0.3, pandoc-1.14.0.2, pandoc-1.14.0.1, pandoc-1.14 (constraint from
user target requires ==3.1.9)
[__0] fail (backjumping, conflict set: pandoc)
After searching the rest of the dependency tree exhaustively, these were the
goals I've had most trouble fulfilling: pandoc
```

* The same rejections but shortened by dropping the common prefix

```
$ cabal build all --enable-tests --enable-benchmarks --dry-run
Resolving dependencies...
Error: [Cabal-7107]
Could not resolve dependencies:
[__0] next goal: pandoc (user goal)
[__0] rejecting: pandoc; 3.1.11, 3.1.10 (constraint from user target requires
==3.1.9)
[__0] rejecting: pandoc; 3.1.9 (constraint from project config
https://stackage.org/lts-21.7/cabal.config requires ==3.0.1)
[__0] rejecting: pandoc; 3.1.8, 3.1.7, 3.1.6.2, 3.1.6.1, 3.1.6, 3.1.5, 3.1.4,
3.1.3, 3.1.2, 3.1.1, 3.1, 3.0.1, 3.0, 2.19.2, 2.19.1, 2.19, 2.18, 2.17.1.1,
2.17.1, 2.17.0.1, 2.17, 2.16.2, 2.16.1, 2.16, 2.15, 2.14.2, 2.14.1, 2.14.0.3,
2.14.0.2, 2.14.0.1, 2.14, 2.13, 2.12, 2.11.4, 2.11.3.2, 2.11.3.1, 2.11.3,
2.11.2, 2.11.1.1, 2.11.1, 2.11.0.4, 2.11.0.2, 2.11.0.1, 2.11, 2.10.1, 2.10,
2.9.2.1, 2.9.2, 2.9.1.1, 2.9.1, 2.9, 2.8.1, 2.8.0.1, 2.8, 2.7.3, 2.7.2, 2.7.1,
2.7, 2.6, 2.5, 2.4, 2.3.1, 2.3, 2.2.3.2, 2.2.3.1, 2.2.3, 2.2.2.1, 2.2.2, 2.2.1,
2.2, 2.1.3, 2.1.2, 2.1.1, 2.1, 2.0.6, 2.0.5, 2.0.4, 2.0.3, 2.0.2, 2.0.1.1,
2.0.1, 2.0.0.1, 2.0, 1.19.2.4, 1.19.2.3, 1.19.2.2, 1.19.2.1, 1.19.2, 1.19.1,
1.19, 1.18, 1.17.2, 1.17.1, 1.17.0.3, 1.17.0.2, 1.17.0.1, 1.17, 1.16.0.2,
1.16.0.1, 1.16, 1.15.2.1, 1.15.2, 1.15.1.1, 1.15.1, 1.15.0.6, 1.15.0.5,
1.15.0.4, 1.15.0.3, 1.15.0.2, 1.14.1, 1.13.2.1, 1.13.2, 1.13.1, 1.13.0.1, 1.13,
1.12.4.2, 1.12.4, 1.12.3.3, 1.12.3.2, 1.12.3.1, 1.12.3, 1.12.2.1, 1.12.2,
1.12.1, 1.12.0.2, 1.12.0.1, 1.12, 1.11.1, 1.11, 1.10.1, 1.10.0.5, 1.10.0.4,
1.10.0.3, 1.10.0.2, 1.10.0.1, 1.10, 1.9.4.5, 1.9.4.4, 1.9.4.3, 1.9.4.2, 1.9.4.1,
1.9.4, 1.9.3, 1.9.2, 1.9.1.2, 1.9.1.1, 1.9.1, 1.9.0.5, 1.9.0.4, 1.9.0.3,
1.9.0.2, 1.9, 1.8.2.1, 1.8.2, 1.8.1.2, 1.8.1.1, 1.8.1, 1.8.0.3, 1.8.0.2,
1.8.0.1, 1.8, 1.6.0.1, 1.6, 1.5.1.1, 1.5.1, 1.5.0.1, 1.5, 1.4, 1.3, 1.2.1, 1.2,
1.1, 1.0.0.1, 1.0, 0.46, 0.45, 0.44, 0.43, 0.42, 0.41, 0.4, 2.11.0.3, 1.15.0.1,
1.15, 1.14.0.4, 1.14.0.3, 1.14.0.2, 1.14.0.1, 1.14 (constraint from user target
requires ==3.1.9)
[__0] fail (backjumping, conflict set: pandoc)
After searching the rest of the dependency tree exhaustively, these were the
goals I've had most trouble fulfilling: pandoc
```

### PR DESCRIPTION
**Pull request by philderbeast**
Fixes #9559. Helps with #4251.

**Template Α: This PR modifies `cabal` behaviour**

Include the following checklist in your PR:

* [x] Patches conform to the [coding conventions](https://github.com/haskell/cabal/blob/master/CONTRIBUTING.md#other-conventions).
* [x] Any changes that could be relevant to users [have been recorded in the changelog](https://github.com/haskell/cabal/blob/master/CONTRIBUTING.md#changelog).
* [ ] ~The documentation has been updated, if necessary.~
* [x] [Manual QA notes](https://github.com/haskell/cabal/blob/master/CONTRIBUTING.md#qa-notes) have been included.
* [x] Tests have been added.

To test, introduce a version bound (or conflict) that is going to reject some versions of a package. Even if this creates an unsolvable combination of versions, that should suffice to trigger the rejection message by the solver for manual testing.

I found no examples of solver rejection messages in the docs.

### CODE REVIEW

**Comment by philderbeast** [comment]
@yvan-sraka I would like to get this in before some formatting changes for showing "constraint from project". This is a subset of your #9541 changes.

**Comment by philderbeast** [comment]
@Mikolaj I had to rerun failed jobs three times to get all green ticks.

image

**Comment by ulysses4ever** [comment]
@philderbeast thanks for the report. The work to prevent this bug is under way in https://github.com/haskell/cabal/pull/9540 If you have energy, your help there would be greatly appreciated because I'm running out of ideas.

**Review by geekosaur — APPROVED** [review]

**Review by yvan-sraka — APPROVED** [review]

**Comment by philderbeast** [comment]
Label `merge+no rebase` is necessary when the pull request is from an organisation.

**Review by ulysses4ever — CHANGES_REQUESTED** [review]
Could we get @grayjay's opinion on it but any chance? She's our solver code owner effectively. Would be great to get her approval on any change to it.

**Review by grayjay — COMMENTED** [review]
@philderbeast Thanks for working on the solver error messages! Please add a note about starting with @yvan-sraka's change to the commit message.

**Comment by philderbeast** [comment]
> Please add a note about starting with @yvan-sraka's change to the commit message.

@yvan-sraka I did not start from your change and sorry I didn't see your work on #4251 earlier. I was working on #9510 and the verbosity of the rejection messages started to annoy me so I raised #9559, saw a quick fix improvement and went for it, found your work and realized I hadn't accounted for hyphenated package names.

**Comment by grayjay** [comment]
> @yvan-sraka I did not start from your change and sorry I didn't see your work on https://github.com/haskell/cabal/issues/4251 earlier. I was working on https://github.com/haskell/cabal/pull/9510 and the verbosity of the rejection messages started to annoy me so I raised https://github.com/haskell/cabal/issues/9559, saw a quick fix improvement and went for it, found your work and realized I hadn't accounted for hyphenated package names.

@philderbeast Sorry, I had misunderstood your comment above. I see that the changes are independent.

File: cabal-testsuite/PackageTests/ShortRejections/pandoc.test.hs
Comment (philderbeast) [reply cabal-testsuite/PackageTests/ShortRejections/pandoc.test.hs]: I got rid of the package test (after confirming it still works). I'm using doctests instead.

Comment (philderbeast) [reply cabal-testsuite/PackageTests/ShortRejections/pandoc.test.hs]: I got rid of the package test (after confirming it still works). I'm using doctests instead.

File: cabal-install-solver/src/Distribution/Solver/Modular/Message.hs
Comment (philderbeast) [reply cabal-install-solver/src/Distribution/Solver/Modular/Message.hs]: Thanks for suggesting this. I much prefer how it looks after this rework.

File: cabal-install-solver/src/Distribution/Solver/Modular/Message.hs
Comment (philderbeast) [reply cabal-install-solver/src/Distribution/Solver/Modular/Message.hs]: I do this now.

File: cabal-testsuite/PackageTests/ShortRejections/pandoc.test.hs
Comment (philderbeast) [reply cabal-testsuite/PackageTests/ShortRejections/pandoc.test.hs:1]: I've added doctests that show this.

**Comment by philderbeast** [comment]
I've left recent commits as-is for review. Please let me know if you'd like some of these squashed.

**Review by grayjay — COMMENTED** [review]
@philderbeast Thanks for the updates and for also handling skipped versions.

> I've left recent commits as-is for review. Please let me know if you'd like some of these squashed.

I would prefer squashing to remove the large test build files before you merge.

File: cabal-install-solver/src/Distribution/Solver/Modular/Message.hs
Comment (philderbeast) [reply cabal-install-solver/src/Distribution/Solver/Modular/Message.hs]: I've done this.

File: cabal-install-solver/src/Distribution/Solver/Modular/Message.hs
Comment (philderbeast) [reply cabal-install-solver/src/Distribution/Solver/Modular/Message.hs]: I've done this and let intercalate check for singleton lists (it does this via calling `intersperse` that uses recursion with `[]` as a first match).

```haskell
build   :: forall a. (forall b. (a -> b -> b) -> b -> b) -> [a]
{-# INLINE [1] build #-}
        -- The INLINE is important, even though build is tiny,
        -- because it prevents [] getting inlined in the version that
        -- appears in the interface file.  If [] *is* inlined, it
        -- won't match with [] appearing in rules in an importing module.
        --
        -- The "1" says to inline in phase 1

build g = g (:) []

concat :: Foldable t => t [a] -> [a]
concat xs = build (\c n -> foldr (\x y -> foldr c y x) n xs)

intersperse :: a -> [a] -> [a]
intersperse _   []      = []
intersperse sep (x:xs)  = x : prependToAll sep xs

intercalate :: [a] -> [[a]] -> [a]
intercalate xs xss = concat (intersperse xs xss)
```

File: cabal-install-solver/src/Distribution/Solver/Modular/Message.hs
Comment (philderbeast) [reply cabal-install-solver/src/Distribution/Solver/Modular/Message.hs:285]: By moving the `reverse` out of `showMessages` and into `showIsOrVs` I'm making it clear that the items are shown in reverse order.

> where the elements are initially added in reverse order

Maybe I'm missing another reversal somewhere but in any case the reversal for presentation is need isn't it?

File: cabal-install-solver/src/Distribution/Solver/Modular/Message.hs
Comment (philderbeast) [reply cabal-install-solver/src/Distribution/Solver/Modular/Message.hs:266]: @grayjay do you know of a test that already uses installed packages? I've seen these listed in stackage `cabal.config` files, such as [stackage/lts-22.4](https://www.stackage.org/lts-22.4/cabal.config) without version numbers:

```
array installed,
base installed,
binary installed,
Cabal installed,
Cabal-syntax installed,
containers installed,
deepseq installed,
...
Win32 installed,
xhtml installed,
```

In this case are you expecting the output to look something like:

`rejecting: base; x.y.z, x.y.z, installed, x.y.z, x.y.z`

File: cabal-install-solver/src/Distribution/Solver/Modular/Message.hs
Comment (philderbeast) [reply cabal-install-solver/src/Distribution/Solver/Modular/Message.hs]: Thanks. `goPSkip` also reverses the list. I added a comment about this and moved the list reversals back into `goPReject` and `goPSkip`.

File: cabal-install-solver/src/Distribution/Solver/Modular/Message.hs
Comment (philderbeast) [reply cabal-install-solver/src/Distribution/Solver/Modular/Message.hs:67]: The module-local version of `showQFNBool` is called twice, once for `Rejecting` and once for `Trying`. Same goes for `showQSNBool`:

```
showQFNBool :: ProgressAction -> QFN -> Bool -> String
showQFNBool a q b = showing a ++ Flag.showQFNBool q b

showQSNBool :: ProgressAction -> QSN -> Bool -> String
showQSNBool a q b = showing a ++ Flag.showQSNBool q b
```

All the functions taking a `ProgressAction` are together and even though the above two functions don't buy us much, they do avoid the call to `showing` and one call to `(++)` being made inline within `showMessages/go`.

File: cabal-install-solver/src/Distribution/Solver/Modular/Message.hs
Comment (philderbeast) [reply cabal-install-solver/src/Distribution/Solver/Modular/Message.hs]: I've renamed the local `show*` functions to `blurb*` and added `blurbOption` so that I don't have to create a singleton list to call `blurbOptions`.

File: cabal-install-solver/src/Distribution/Solver/Modular/Message.hs
Comment (philderbeast) [reply cabal-install-solver/src/Distribution/Solver/Modular/Message.hs]: Thanks for the explanation. I made the change you suggested.

File: cabal-install-solver/src/Distribution/Solver/Modular/Message.hs
Comment (philderbeast) [reply cabal-install-solver/src/Distribution/Solver/Modular/Message.hs:252]: The two functions with doctests I would say are a stable interface, they do contribute to the logged output of the solver. As unit tests, they are smaller units.

I've encountered various problems with doctests over the years but haven't had any problems so far with these in this project. Here's how I run doctests:

```
$ make doctest-install
$ cabal repl --with-ghc=doctest --repl-options="-w" --project-file=cabal.project.doctest cabal-install-solver
```

A bonus of doctests (stating the obvious) is that they also serve as documentation.

I don't think failing doctests could be any easier to modify, the failure report has the actual output. This can be copied from the report and pasted directly as the expectation.

**Comment by Mikolaj** [comment]
Let me take the liberty and rebase on master that just got a fix to CI.

**Comment by Mikolaj** [comment]
@mergify rebase

File: cabal-install-solver/src/Distribution/Solver/Modular/Message.hs
Comment (grayjay) [reply cabal-install-solver/src/Distribution/Solver/Modular/Message.hs:252]: I have no problem with using doctests (as long as they are added to the CI), but I think that `tryVs` and `showIsOrVs` shouldn't be tested directly because they don't form a stable interface. I thought about it some more, and the reason that it seems unstable is that it exposes the new `IsOrVs` type, which seems very tightly tied to the current logic for displaying rejected package options. If someone made a further improvement to the format for displaying rejected versions, it is very likely that they would modify or even remove the `IsOrVs` type.

Another idea for testing a more stable interface would be to make a new function that calls both `tryVs` and `showIsOrVs` and add the doctests to it. The function would convert a `QPN` and a `[POption]` to a string, which I think is functionality that would continue to be needed through a wide variety of potential improvements. Another benefit of testing at that level is that when contributors make additional improvements to the format of the rejecting message, they can simply update the expected output for the existing tests (in addition to adding new tests), and the diff will clearly show how the format was improved.

**Comment by Mikolaj** [comment]
@ulysses4ever: I think your change request is holding up the merge. Are you satisfied? @philderbeast: anything you'd like to add in that review thread?

**Comment by Mikolaj** [comment]
@philderbeast: BTW, if you'd like to hold the merge until you finished the other discussions, please kindly remove the merge labels.

**Comment by ulysses4ever** [comment]
@Mikolaj I wanted to see how the discussions between @philderbeast and @grayjay resolve.

**Comment by Mikolaj** [comment]
Right, I missed that in my CI-breakage panic mode. :)

File: cabal-install-solver/src/Distribution/Solver/Modular/Message.hs
Comment (philderbeast) [reply cabal-install-solver/src/Distribution/Solver/Modular/Message.hs:252]: I've added doctests in `cabal-install-solver` to the Makefile `doctest` target recipe so they'll run in CI.

We don't export functions `tryVs`, `showIsOrVs` but can test them with doctests the way we have it set up.

File: cabal-install-solver/src/Distribution/Solver/Modular/Message.hs
Comment (philderbeast) [reply cabal-install-solver/src/Distribution/Solver/Modular/Message.hs:272]: I've added the following unit tests that test skipping:

```
$ cabal run cabal-install:unit-tests -- -p "5960"
...
Unit Tests
  UnitTests.Distribution.Solver.Modular.Solver
    Solver log
      package versions abbreviation (issue #5960)
        show skipping singleton:                OK
        show skipping singleton, installed:     OK
        show skipping versions list:            OK
        show skipping versions list, installed: OK

All 4 tests passed (0.00s)
```

I couldn't see how to trigger multiple rejections like in the pandoc example of #9559.

Comment (philderbeast) [reply cabal-install-solver/src/Distribution/Solver/Modular/Message.hs:272]: > I don't think that it's necessary to exclude installed packages from the new formatting, if you are interested in handling them, too.

We shorten the rejecting and skipping messages a lot by putting the package name in front of a list of versions and not repeating it hyphenated with the versions. I tried the same thing with installed packages and thought it harder to read and not so effective at shortening. Here is the test example I tried:

```
[__0] trying: B-1.0.0 (user goal)
[__1] next goal: A (dependency of B)
[__1] rejecting: A; 3.0.0/installed-3.0.0 (conflict: B => A==4.0.0)
[__1] skipping: A; 2.0.0/installed-2.0.0, 1.0.0/installed-1.0.0 ...
[__1] fail (backjumping, conflict set: A, B)
```

I would rather not extend this pull request to include abbreviating lists that include installed packages.

Comment (philderbeast) [reply cabal-install-solver/src/Distribution/Solver/Modular/Message.hs:272]: @grayjay should I bring back the `PackageTests/ShortRejections` tests? They were able to trigger multiple rejections like in the pandoc example of #9559.

Comment (grayjay) [reply cabal-install-solver/src/Distribution/Solver/Modular/Message.hs:272]: Thanks for adding the new unit tests. I think that those are enough, and this change doesn't need the package test. I'm guessing that packages are rejected rather than skipped when there is a conflict due to a global constraint.

**Review by grayjay — APPROVED** [review]

File: cabal-install/tests/UnitTests/Distribution/Solver/Modular/Solver.hs
Comment (philderbeast) [reply cabal-install/tests/UnitTests/Distribution/Solver/Modular/Solver.hs]: Thanks for the suggestion, I've done this.

**Comment by philderbeast** [comment]
@ulysses4ever this is ready for your approval.

**Review by ulysses4ever — APPROVED** [review]
Perfect, thanks!

**Comment by philderbeast** [comment]
Label merge+no rebase is necessary when the pull request is from an organisation.

**Comment by ulysses4ever** [comment]
Please, try not to forget squashing before merging.

**Comment by philderbeast** [comment]
@ulysses4ever a momentary lapse of diligence on my part, sorry about that.

**Pull request merged**
